### PR TITLE
[FLINK-27995][table] Upgrade Janino version to 3.1.9

### PR DESCRIPTION
**Pull request by snuyanzin**
## What is the purpose of the change

This is a draft version of Janino's upgrade
In https://github.com/apache/flink/pull/21203 Calcite was updated to 1.28.0
However Janino was still 3.0.11
The reason is that there is still a list of issues on Janino 3.1.x for the fine working cases with Janino 3.0.11
https://github.com/janino-compiler/janino/issues/185
https://github.com/janino-compiler/janino/issues/187
https://github.com/janino-compiler/janino/issues/188

Currently it seems that now only one(https://github.com/janino-compiler/janino/issues/188) is not fixed (looks like it is fixed partially in master)
There is a workaround for that at `org.apache.flink.table.planner.codegen.CodeGenUtils#genToExternalConverterAllWithLegacy` (hope it will be fixed soon and workaround will be removed)

UPD: a follow up jira issue to remove WA https://issues.apache.org/jira/browse/FLINK-30984

## Verifying this change

It looks like it should be covered by tests in flink-table-planner

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes )
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: ( no)
  - The serializers: ( no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: ( no )
  - The S3 file system connector: ( no)

## Documentation

  - Does this pull request introduce a new feature? ( no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by snuyanzin** [comment]
@flinkbot run azure

**Comment by snuyanzin** [comment]
@flinkbot run azure

**Comment by snuyanzin** [comment]
@fsk119  since you are a reporter of FLINK-27995
could you have a look here please?

**Comment by fsk119** [comment]
@snuyanzin I am not familiar with janino... I think you may need to find others to review this.

**Review by lincoln-lil — APPROVED** [review]
@snuyanzin LGTM! 
We can benefit from the new janino version, I think this change(the only 
 temporary workaround for the null cast issue) is fine with me.

**Comment by lincoln-lil** [comment]
@twalthr do you have any concern on this pr?

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala [hunk 1]
```diff
@@ -1020,16 +1021,19 @@ object CodeGenUtils {
     }
 
     // convert internal format to target type
-    val externalResultTerm = if (isInternalClass(targetDataType)) {
-      s"($targetTypeTerm) ${internalExpr.resultTerm}"
+    val (externalResultTerm, resultType) = if (isInternalClass(targetDataType)) {
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:1024]: can we use more meaningful names here? `resultType` does not improve code readability. `externalResultTypeTerm` is the right name here.

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala [hunk 2]
```diff
@@ -922,7 +922,8 @@ object CodeGenUtils {
     if (targetDataType.getConversionClass.isPrimitive) {
       externalResultTerm
     } else {
-      s"${internalExpr.nullTerm} ? null : ($externalResultTerm)"
+      // Cast of null is required because of janino issue https://github.com/janino-compiler/janino/issues/188
+      s"${internalExpr.nullTerm} ? (${typeTerm(targetDataType.getConversionClass)}) null : ($externalResultTerm)"
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/codegen/CodeGenUtils.scala:926]: introduce a variable `externalResultTypeTerm = typeTerm(targetDataType.getConversionClass)`

**Review by twalthr — APPROVED** [review]
LGTM I just have some suggestions to improve the readability of these complex code gen pieces.

**Comment by snuyanzin** [comment]
Rebased to resolve ci issue
```
Reading package lists...
E: Unsupported file ./libssl1.0.0_1.0.2n-1ubuntu5.10_amd64.deb given on commandline
##[error]Bash exited with code '100'.
```
https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=46553&view=logs&j=bea52777-eaf8-5663-8482-18fbc3630e81&t=11b0df07-3e5e-58da-eb81-03003e470195&l=1836

**Pull request merged**